src/ydata_synthetic/utils/gumbel_softmax.py

```diff
@@ -0,0 +1,66 @@
+"""Gumbel-Softmax layer implementation.
+Reference: https://arxiv.org/pdf/1611.04051.pdf"""
+from typing import Optional
+
+# pylint: disable=E0401
+from tensorflow import Tensor, TensorShape, concat, one_hot, split, squeeze, stop_gradient
+from tensorflow.keras.layers import Activation, Layer
+from tensorflow.math import log
+from tensorflow.nn import softmax
+from tensorflow.random import categorical, uniform
+
+from ydata_synthetic.preprocessing.base_processor import BaseProcessor
+
+TOL = 1e-20
+
+
+def gumbel_noise(shape: TensorShape) -> Tensor:
+    """Create a single sample from the standard (loc = 0, scale = 1) Gumbel distribution."""
+    uniform_sample = uniform(shape)
+    return -log(-log(uniform_sample + TOL) + TOL)
+
+
+class GumbelSoftmaxLayer(Layer):
+    "A Gumbel-Softmax layer implementation that should be stacked on top of a categorical feature logits."
+
+    def __init__(self, tau: float = 0.2, name: Optional[str] = None):
+        super().__init__(name = name)
+        self.tau = tau
+
+    # pylint: disable=W0221, E1120
+    def call(self, _input):
+        """Computes Gumbel-Softmax for the logits output of a particular categorical feature."""
+        noised_input = _input + gumbel_noise(_input.shape)
+        soft_sample = softmax(noised_input/self.tau, -1)
+        hard_sample = stop_gradient(squeeze(one_hot(categorical(log(soft_sample), 1), _input.shape[-1]), 1))
+        return hard_sample, soft_sample
+
+
+class ActivationInterface(Layer):
+    """An interface layer connecting different parts of an incoming tensor to adequate activation functions.
+    The tensor parts are qualified according to the passed processor object.
+    Processed categorical features are sent to specific Gumbel-Softmax layers.
+    Processed features of different kind are sent to a TanH activation.
+    Finally all output parts are concatenated and returned in the same order.
+
+    The parts of an incoming tensor are qualified by leveraging a data processor object (from the synthesizer)."""
+
+    def __init__(self, processor: BaseProcessor, name: Optional[str] = None):
```

Comment:
Why receiving the processor? 

Please assume or create a metadata kind of object where this kind of information can be stored.

Reply:
Added a metadata object to the processor to be leveraged here and in this kind of developments!